### ISSUE TITLE
heatmap_slicer indexing is off by one

### ISSUE THREAD

**Issue by drfeinberg** [ianhi/mpl-interactions#88]
I was wondering if someone could give me a hand with heatmap_slicer(). I have a pcolormesh plot that I'd like to be able to view slices of using the mouse. It's got X, Y, and Z coordinates. X and Y are 1D numpy arrays and the Z axis is a 2D array. I'm stumped as to what to put in the heatmaps keyword argument. It's a spectrogram.

**Comment by ianhi** [comment]
Hi @drfeinberg the `heatmaps` argument is somewhat equivalent to the `C` argument of  of https://matplotlib.org/api/_as_gen/matplotlib.pyplot.pcolormesh.html

It sounds like in your case you should be able to do:
```python
fig, axes = heatmap_slicer(X, Y, Z)
```

If you can provide a complete minimal example then that will make it easier to create an example of what to do. Also, one thing that may be worthwhile is starting with the example in docs and then changing it piece by piece to match your data https://mpl-interactions.readthedocs.io/en/latest/examples/heatmap-slicer.html

**Comment by ianhi** [comment]
Also, if this was unclear to you it's probably unclear to someone else, so if at any point you have a suggestion for another name for that argument I would love to hear it :)

**Comment by ianhi** [comment]
To follow up on my dangling "somewhat equivalent":

The array(s) you pass to the heatmaps argument are directly passed to the pcolormesh `C` argument. If you pass a single 2D array then it will be passed to pcolormesh, if you pass an array with shape `(n, len(X), len(Y)` then there will be `n` calls to pcolormesh creating `n` heatmaps to compare. 

If it helps the relevant lines of code are:
https://github.com/ianhi/mpl-interactions/blob/ef118f70a998b9d72108b4514a4d2ed9d803349a/mpl_interactions/generic.py#L77-L85
and 
https://github.com/ianhi/mpl-interactions/blob/ef118f70a998b9d72108b4514a4d2ed9d803349a/mpl_interactions/generic.py#L95-L96

**Comment by drfeinberg** [comment]
Gotcha.  It looks like my problem is uneven array sizes.  Thanks for the help here.   I'll play around and see if I can get this going.

**Comment by drfeinberg** [comment]
As far as I can tell, the error is due to X being larger than Y.  X is time, Y is frequency, Z are amplitude bins.  They're all different shapes.  I'm not sure how I'd reshape the data to fix this.  Any ideas?

np.shape(X) gets (696,)  
np.shape(Y) gets (313,)
np.shape(Z) gets (312, 695)

Here's my code, spectrogram generated using the parselmouth-praat package...
`
sound = parselmouth.Sound("/home/david/good_sound.wav")
sound.pre_emphasize()
spectrogram = sound.to_spectrogram(window_length=0.05, maximum_frequency=5500)
X, Y = spectrogram.x_grid(), spectrogram.y_grid()
Z = 10 * np.log10(spectrogram.values)
print(X.shape[0])
print(Y.shape[0])
print(Z.shape[0])
dynamic_range=70
vmin = sg_db.max() - dynamic_range
#plot = plt.pcolormesh(X, Y, Z, vmin, cmap='afmhot')
fig, axes = heatmap_slicer(X, Y, Z, slices='both',
                        heatmap_names=('Spectrogram', 'Frequency', 'Time'),
                        labels=('Time','Frequency'),
                        interaction_type='move')
`

![error](https://user-images.githubusercontent.com/20147038/91196133-bf43e680-e6c7-11ea-8e17-16b839b863f3.png)

**Comment by ianhi** [comment]
oooh hmm. You seem to have found a bug with `heatmap_slicer`. Which I think makes you the first bugfinder for this library!! 
Thanks for the code and shapes it helped narrow this down.

Minimal example:
```python
%matplotlib widget
import matplotlib.pyplot as plt
import numpy as np
from mpl_interactions import heatmap_slicer
X = np.linspace(0, np.pi, 696)
Y = np.linspace(0, 10, 313)
Z = np.random.randn(312, 695)
plt.figure()
plt.pcolormesh(X, Y, Z) # works
heatmap_slicer(X, Y, Z) # breaks
```
(n.b. you can format code in a github issue and add syntax highlighting by using triple backticks see https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks)

So what's happening is that I've been using pcolormesh slightly wrong for years
From the [docs ](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.pcolormesh.html#axes-pcolormesh-grid-orientation)
> The dimensions of X and Y should be one greater than those of C. Alternatively, X, Y and C may have equal dimensions, in which case the last row and column of C will be ignored.

I've always used it with the same shape and just not realized I was dropping the laster column of C. So the pcolormesh call succeeds but the logic that creates a line was not working correctly. I will try to fix this tonight or tomorrow but in the meantime you can just shorten your X and Y arrays by 1 as a workaround

```python
heatmap_slicer(X[:-1],Y[:-1],Z)
```

**Comment by drfeinberg** [comment]
Thanks!  I really appreciate you having a look.  I was sure I just didn't really understand what was going on, but I feel a little better knowing it was a bug.  Thanks for the workaround too.  Good luck with the fix!

**Comment by ianhi** [comment]
One other consequence of my prior misinterpretation of how heatmapslicer works is that the line never intersects the middle of pixels, instead it snaps to the rows between pixels:
![hmap-slicer](https://user-images.githubusercontent.com/10111092/91205873-426b3980-e6d4-11ea-846c-6a2a920f0d43.png)

**Comment by drfeinberg** [comment]
Oh wow.  That's interesting.  That might give some unexpected values, eh?

**Comment by ianhi** [comment]
Out of curiosity what would you expect to be used as the X axis for the plot of the slices? Currently, I just use the `X` array, but reading the pcolormesh docs more carefully it seems that I should really be interpreting the X coordinates as pixel edges. So I maybe it should use the coordinates at the center of each pixel. 

I'm thinking that in your case I if I were to plot the 0th horizontal slice of z I should transform X like this for the plot:
```python
x = (X[1:] + X[:-1])/2
plt.plot(x, Z[0,:])
```

**Comment by ianhi** [comment]
I want to make sure I get this "right" as in do what's expected. But I've only ever used this for quick comparisons so I'm not 100% confident on what "right" is.

**Comment by drfeinberg** [comment]
Yeah... that's why I was thinking you'd get unexpected values.  AFAIK, and I'm not an expert, is just what you said.  When I run an analysis using the X axis, I normally take the average (the middle) of the edges.  I would definitely check with someone else though because I'm not super confident here.

**Comment by drfeinberg** [comment]
Huzzah!!! Success!



![Interactive Spectrogram](https://user-images.githubusercontent.com/20147038/91210063-a8f35600-e6da-11ea-86a7-fc810b172ea9.gif)

**Comment by ianhi** [comment]
:tada: :tada:  looks great!

I think the usage with sound data is really cool. If you ever think of another interaction you'd like please don't hesitate to open an issue proposing it! So far this library is all functions that I've found a need for in my research but I think it would be awesome to have other sources of inspiration as well :)

**Comment by ianhi** [comment]
Thanks for the report @drfeinberg! This is now fixed and has published to pypi. `pip install mpl-interactions==0.5.2`

I learned quite a lot about pcolormesh along the way and actually found another bug with heatmap slicer so overall a productive enterprise! One thing to note if you plan on using pcolormesh is that it had some significant but hard to notice improvements between matplotlib 3.2 and 3.3
- https://matplotlib.org/3.3.1/users/whats_new.html#pcolor-and-pcolormesh-now-accept-shading-nearest-and-auto
- https://matplotlib.org/3.3.1/gallery/images_contours_and_fields/pcolormesh_grids.html#sphx-glr-gallery-images-contours-and-fields-pcolormesh-grids-py

so it may be worth your while to upgrade matplotlib as well. Though I wrote fixes in such a way that they should also work for earlier versions :crossed_fingers:.
